### DISCUSSION TITLE
Losing the xml tag when sending document

### DISCUSSION THREAD

**Discussion by yvansteene** [phax/phase4#210] [Q&A] [answered]
Hello, when sending an XML, we lose the <xml> tag at the beginning of the document, which causes issues for some of our receiver.

Any idea where the problem can be?

**Comment by phax** [comment]
Can you please be a more specific what "Losing" means in this context?
My assumption is, that you are losing the XML namespace declarations (`xmlns`).
Is that correct?

**Reply by yvansteene** [reply]
It's <?xml version="1.0" encoding="UTF-8"?> tag at the beginning of the document.
We don't have it in the documents we receive either.

**Reply by phax** [reply]
Okay, then I need more information please. Is it in Peppol or somewhere else? What setup are you using? Feel free to email me with more details (see https://github.com/phax for the address), if you don't want to share them here

**Reply by L3Mars** [reply]
Hello, i am adding more information to @yvansteene question.

we are sending Peppol ubl documents using Phase4. Our issue is that during document transmission, it seems that the <?xml version=\"1.0\" encoding=\"UTF-8\"?> tag is missing, even though it was present in the original XML. Here is an example of the document transmission code.

```java
final Document aDocument = DOMReader.readXMLDOM(ubl.getXml());
final Element aPayloadElement = aDocument.getDocumentElement();

Phase4PeppolSender.builder()
				.messageID(...)				.documentTypeID(Phase4PeppolSender.IF.createDocumentTypeIdentifierWithDefaultScheme(ublHeader.getTypeDocumentId()))
				.processID(Phase4PeppolSender.IF.createProcessIdentifierWithDefaultScheme(ublHeader.getProcessId()))
				.senderParticipantID(Phase4PeppolSender.IF.createParticipantIdentifierWithDefaultScheme(ubl.getSenderId()))
				.receiverParticipantID(aReceiverID)
				.senderPartyID(this.partyId)
				.payload(aPayloadElement)
				.smpClient(new SMPClientReadOnly(Phase4PeppolSender.URL_PROVIDER, aReceiverID, smlService.getSML()).setSecureValidation(false))
				.rawResponseConsumer(new AS4RawResponseConsumerSQL())
				.sendMessageAndCheckForReceipt(exceptionConsumer);
```

When receiving a document, we use this

```java
String document = serializer.writeToString(m_aPeppolStandardBusinessDocumentHeader.getBusinessMessageNoClone());
```

Then we simply save our UBL in an S3 bucket.

**Reply by phax** [reply]
@L3Mars  thanks for the clarification - that helps :)
* Sending code looks okay. Have you verified, that `aPayloadElement` is indeed a UBL `Invoice` element (or comparable)?
* The receiving code also looks good - except that I of course don't know what `serializer.writeToString` does. Have you checked if `getBusinessMessageNoClone()` returns an `Invoice` element as expected?

Note: the `toString` method of the DOM elements is dangerous, because it may show "null" even if it is valid. You need to inspect the fields for local name and namespace URI!

**Reply by L3Mars** [reply, marked as answer]
Thanks ! we are so dumb.. 

when receiving document, there's is a specific configuration of the serializer for remove xml declaration

```
		DOMImplementationLS lsImpl = (DOMImplementationLS) ublDocument.getOwnerDocument().getImplementation().getFeature("LS", "3.0");
		LSSerializer serializer = lsImpl.createLSSerializer();
		serializer.getDomConfig().setParameter("xml-declaration", false);
```
Removing the last line fix the problem.                                               
Sorry for the inconvenience.

**Reply by phax** [reply]
LOL - great that you were able to resolve it. The first sewat beads started to appear ;-)
